client/components/combobox.jsx

```diff
@@ -1,128 +1,166 @@
-				onClick=     {this.props.trigger == 'click' ? ()=>{this.handleDropdown(true);} : undefined}>
-				<input
-					type='text'
-					onChange={(e)=>this.handleInput(e)}
-					value={this.state.value || ''}
-					placeholder={this.props.placeholder}
-					onBlur={(e)=>{
-						if(!e.target.checkValidity()){
-							this.setState({
-								value : this.props.default
-							}, ()=>this.props.onEntry(e));
-						}
-					}}
-				/>
-			</div>
-		);
-	},
-	renderDropdown : function(dropdownChildren){
-		if(!this.state.showDropdown) return null;
-		if(this.props.autoSuggest && !this.state.inputFocused){
-			const suggestMethod = this.props.autoSuggest.suggestMethod;
-			const filterOn = _.isString(this.props.autoSuggest.filterOn) ? [this.props.autoSuggest.filterOn] : this.props.autoSuggest.filterOn;
-			const filteredArrays = filterOn.map((attr)=>{
-				const children = dropdownChildren.filter((item)=>{
-					if(suggestMethod === 'includes'){
-						return item.props[attr]?.toLowerCase().includes(this.state.value.toLowerCase());
-					} else if(suggestMethod === 'startsWith'){
-						return item.props[attr]?.toLowerCase().startsWith(this.state.value.toLowerCase());
-					}
+		setFilteredOptions(filtered);
+	};
+
+	/* eslint-disable brace-style */
+
+	// Handle keyboard navigation
+	const handleKeyDown = (evt)=>{
+		const modifiers = ['Meta', 'Shift', 'Alt', 'Control', 'Tab'];
+		if(inputFocused || (currentOption >= 0)) {
+			const optionsLength = filteredOptions.length;
+
+			if((evt.key === ' ') && (inputValue === '')){
+				evt.preventDefault();
+				setShowDropdown(!showDropdown);
+			}
+
+			// ArrowDown moves to the next option
+			else if(evt.key === 'ArrowDown') {
+				evt.preventDefault();
+				if((currentOption === -1) && (showDropdown === false)){
+					setShowDropdown(true);
+				};
+				const nextIndex = currentOption + 1;
+				if(nextIndex < optionsLength) {
+					setCurrentOption(nextIndex);
+				} else {
+					setCurrentOption(0);
+				}
+			}
+
+			// ArrowUp moves to the previous option
+			else if(evt.key === 'ArrowUp') {
+				evt.preventDefault();
+				const prevIndex = currentOption - 1;
+				if(prevIndex >= 0) {
+					setCurrentOption(prevIndex);
+				} else {
+					setCurrentOption(-1);
+					inputRef.current.focus();
+				}
+			}
+
+			// Escape key closes the dropdown
+			else if(evt.key === 'Escape'){
+				setCurrentOption(-1);
+				inputRef.current.focus();
+				setShowDropdown(false);
+			}
+
+			// Backspace key while menu is open still deletes characters in input
+			else if((evt.key === 'Backspace') && showDropdown){
+				inputRef.current.focus();
+			}
+
+			else if((evt.key === 'Tab')){
+				setCurrentOption(-1);
+				setShowDropdown(false);
+			}
+
+			// Prevent modifier keys from triggering dropdown (example, shift+tab or tab to move focus)
+			else if(!modifiers.includes(evt.key)) {
+				setShowDropdown(true);
+			}
+		}
+
+
+	};
```

Comment:
With that many else ifs, we might benefit from using the hated switch:

```suggestion
    const handleKeyDown = (evt) => {
    const modifiers = ['Meta', 'Shift', 'Alt', 'Control', 'Tab'];
    const isArrowDown = evt.key === 'ArrowDown';
    const isArrowUp = evt.key === 'ArrowUp';
    const optionsLength = filteredOptions.length;

    if (inputFocused || currentOption >= 0) {
        switch (evt.key) {
            case ' ':
                if (inputValue === '') {
                    evt.preventDefault();
                    setShowDropdown(prev => !prev);
                }
                break;
                
                // ArrowDown moves to the next option
            case 'ArrowDown':
                evt.preventDefault();
                setShowDropdown(true);
                setCurrentOption((currentOption + 1) % optionsLength);
                break;
                
                // ArrowUp moves to the previous option
            case 'ArrowUp':
                evt.preventDefault();
                const newIndex = currentOption - 1;
                setCurrentOption(newIndex >= 0 ? newIndex : -1);
                if (newIndex === -1) inputRef.current.focus();
                break;
                
                // Escape key closes the dropdown
            case 'Escape':
                setShowDropdown(false);
                setCurrentOption(-1);
                inputRef.current.focus();
                break;
                
                // Backspace key while menu is open still deletes characters in input
            case 'Backspace':
                if (showDropdown) inputRef.current.focus();
                break;
            case 'Tab':
                setShowDropdown(false);
                setCurrentOption(-1);
                break;
                
                // Prevent modifier keys from triggering dropdown (example, shift+tab or tab to move focus)
            default:
                if (!modifiers.includes(evt.key)) setShowDropdown(true);
                break;
        }
    }
};

```

Please double check it in case i made many errors.

Reply:
I considered it, but I opted for else if before I personally think it is more readable and understandable (don’t have to worry about “making many errors”. It doesn’t save many (any?) lives, either (though I can see I have some extra white space). 

I’ll reconsider if Switch/Case has a huge amount of support among reviewers

Reply:
I'm going to mark this as 'resolved' for right now mostly because it's enormous amount of text in the middle of the thread.  If anyone else comments that they want Switch/Case statements rather than else/if, I'll *switch* it.

Reply:
I'll add another vote for `switch`. This seems like the perfect use case for that.